src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryCopy.java

```diff
@@ -0,0 +1,574 @@
+     * @param srcOffsetBytes the starting offset, in bytes, of the source segment.
+     * @param dstArray the destination byte array.
+     * @param dstIndexBytes the starting index of the destination byte array.
+     * @param dstCopyLengthBytes the number of byte elements to be copied.
+     */
+    @ForceInline
+    public static void copyToArray(
+            MemorySegment srcSegment, long srcOffsetBytes,
+            byte[] dstArray, int dstIndexBytes, int dstCopyLengthBytes) {
+        Objects.requireNonNull(srcSegment);
+        MemorySegment srcSegmentSlice = srcSegment.asSlice(srcOffsetBytes, dstCopyLengthBytes);
+        MemorySegment dstSegmentSlice = MemorySegment.ofArray(dstArray).asSlice(dstIndexBytes, dstCopyLengthBytes);
+        dstSegmentSlice.copyFrom(srcSegmentSlice);
+    }
+
+    //CHAR
+    /**
+     * Copies a number of char elements from a source char array to a destination segment,
+     * starting at a given array index, and a given segment offset (expressed in bytes).
+     * @param srcArray the source char array.
+     * @param srcIndexChars the starting index of the source char array.
+     * @param srcCopyLengthChars the number of char elements to be copied.
+     * @param dstSegment the destination segment.
+     * @param dstOffsetBytes the starting offset, in bytes, of the destination segment.
+     */
+    @ForceInline
+    public static void copyFromArray(
+            char[] srcArray, int srcIndexChars, int srcCopyLengthChars,
+            MemorySegment dstSegment, long dstOffsetBytes) {
+        copyFromArray(srcArray, srcIndexChars, srcCopyLengthChars, dstSegment,dstOffsetBytes, ByteOrder.nativeOrder());
+    }
+
+    /**
+     * Copies a number of char elements from a source char array to a destination segment,
+     * starting at a given array index, and a given segment offset (expressed in bytes), using the given byte order.
+     * @param srcArray the source char array.
+     * @param srcIndexChars the starting index of the source char array.
+     * @param srcCopyLengthChars the number of char elements to be copied.
+     * @param dstSegment the destination segment.
+     * @param dstOffsetBytes the starting offset, in bytes, of the destination segment.
+     * @param order the byte order to be used for the copy operation. If the specified byte order is
+     * different from the native order, a byte swap operation will be performed on each array element.
+     */
+    @ForceInline
+    public static void copyFromArray(
+            char[] srcArray, int srcIndexChars, int srcCopyLengthChars,
+            MemorySegment dstSegment, long dstOffsetBytes,
+            ByteOrder order) {
+        Objects.requireNonNull(dstSegment);
+        MemorySegment srcSegmentSlice =
+                MemorySegment.ofArray(srcArray).asSlice(srcIndexChars * 2L, srcCopyLengthChars * 2L);
+        MemorySegment dstSegmentSlice = dstSegment.asSlice(dstOffsetBytes, srcCopyLengthChars * 2L);
```

Comment:
The variable name `srcCopyLengthChars` is for clarity to the user of the API, who does not see the implementation code. Because there are two different types of units in the method call, one in Chars and one in Bytes, it is important to clarify the units of the measure as well as which of source or destination it is associated with.

Reply:
Honestly, I tend to agree with Jorn here. In a copy, there is:
* a source (either an array or a segment)
* a source offset (expressed either as element index, or as byte offset)
* a destination (either an array or a segment)
* a destination offset (expressed either as element index, or as byte offset)
* a length (expressed either as number of elements, or as number of bytes)

There's only one length - it's important to distinguish between Chars vs. Bytes - e.g. copyLengthChars vs. copyLengthBytes. But adding the `src` seems confusing because the `length` parameter is, IMHO neutral, and just indicates how many "things" have to be bulk-copied.

In fact, I'll go further and say that, when editing this patch I kept finding the placement of the length argument confusing. Having it as the trailing argument would be clearer IMHO.